app/coffee/RedisManager.coffee

```diff
@@ -56,11 +61,19 @@ module.exports = RedisManager =
 		multi = rclient.multi()
 		multi.get keys.docLines(doc_id:doc_id)
 		multi.get keys.docVersion(doc_id:doc_id)
+		multi.get keys.docHash(doc_id:doc_id)
 		multi.get keys.projectKey(doc_id:doc_id)
 		multi.get keys.ranges(doc_id:doc_id)
-		multi.exec (error, [docLines, version, doc_project_id, ranges])->
+		multi.exec (error, [docLines, version, storedHash, doc_project_id, ranges])->
 			timer.done()
 			return callback(error) if error?
+
+			# check sha1 hash value if present
+			if docLines? and storedHash?
```

Comment:
This logic should have some unit tests I think. It's non-trivial enough, and difficult to test manually.